core/common/src/ByteStringExt.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2010-2023 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.io
+
+import kotlinx.io.bytestring.ByteString
+import kotlinx.io.bytestring.indices
+import kotlinx.io.bytestring.isEmpty
+import kotlinx.io.bytestring.unsafe.UnsafeByteStringApi
+import kotlinx.io.bytestring.unsafe.UnsafeByteStringOperations
+import kotlin.math.min
+
+/**
+ * Writes subsequence of data from [byteString] starting at [startIndex] and ending at [endIndex] into a sink.
+ *
+ * @param byteString the byte string whose subsequence should be written to a sink.
+ * @param startIndex the first index (inclusive) to copy data from the [byteString].
+ * @param endIndex the last index (exclusive) to copy data from the [byteString]
+ *
+ * @throws IndexOutOfBoundsException when [startIndex] or [endIndex] is out of range of [byteString] indices.
+ * @throws IllegalArgumentException when `startIndex > endIndex`.
+ * @throws IllegalStateException if the sink is closed.
+ */
+@OptIn(DelicateIoApi::class)
+public fun Sink.write(byteString: ByteString, startIndex: Int = 0, endIndex: Int = byteString.size) {
```

Comment:
I can easily imagine how someone can use `readTo(ByteStringBuilder)`, but not sure how practical it would be.
So I would abstain from adding such extensions until there will be some convincing use cases.